packages/flutter/lib/src/painting/matrix_utils.dart

```diff
@@ -17,23 +17,15 @@ abstract final class MatrixUtils {
   ///
   /// Otherwise, returns null.
   static Offset? getAsTranslation(Matrix4 transform) {
```

Comment:
These matrix checks run a lot. even though its ugly, I'd prefer we keep the older but definitely faster approach for these.

Reply:
Fair point! Will revert.

Reply:
It'd be great to understand if/how we COULD make the new bits as fast as the old bits.

Reply:
I was curious as to exactly how these methods compare, so I went ahead and ran some benchmarks.

<details> <summary>source code</summary>

This was my Dart file for "`isIdentity`, original setup":

```dart
import 'dart:math' as math;
import 'package:benchmark_harness/benchmark_harness.dart';
import 'package:vector_math/vector_math_64.dart';

final rng = math.Random();
const identity = <double>[
  1.0, 0.0, 0.0, 0.0, //
  0.0, 1.0, 0.0, 0.0, //
  0.0, 0.0, 1.0, 0.0, //
  0.0, 0.0, 0.0, 1.0, //
];
List<double> get randomValues {
  final randomCount = rng.nextInt(17);
  final x = [
    for (int i = 0; i < 16; i++)
      if (i < randomCount) (rng.nextDouble() - 0.5) * math.pow(2, rng.nextInt(10)) else null
  ]..shuffle();
  return [for (final (index, item) in x.indexed) item ?? identity[index]];
}

final List<Matrix4> data = [
  for (int i = 0; i < 1e5; i++)
    if (rng.nextBool()) Matrix4.identity() else Matrix4.fromList(randomValues),
];

abstract class MatrixBenchmark extends BenchmarkBase {
  MatrixBenchmark(super.name);

  bool isIdentity(Matrix4 a);

  @override
  void run() {
    for (final Matrix4 a in data) {
      final _ = isIdentity(a);
    }
  }
}

class EqualityChain extends MatrixBenchmark {
  EqualityChain() : super('equality chain');

  @override
  bool isIdentity(Matrix4 a) {
    return a.storage[0] == 1.0 &&
        a.storage[1] == 0.0 &&
        a.storage[2] == 0.0 &&
        a.storage[3] == 0.0 &&
        a.storage[4] == 0.0 &&
        a.storage[5] == 1.0 &&
        a.storage[6] == 0.0 &&
        a.storage[7] == 0.0 &&
        a.storage[8] == 0.0 &&
        a.storage[9] == 0.0 &&
        a.storage[10] == 1.0 &&
        a.storage[11] == 0.0 &&
        a.storage[12] == 0.0 &&
        a.storage[13] == 0.0 &&
        a.storage[14] == 0.0 &&
        a.storage[15] == 1.0;
  }
}

class PatternMatching extends MatrixBenchmark {
  PatternMatching() : super('pattern-matching');

  @override
  bool isIdentity(Matrix4 a) {
    if (a.storage
        case [
          1.0, 0.0, 0.0, 0.0, //
          0.0, 1.0, 0.0, 0.0, //
          0.0, 0.0, 1.0, 0.0, //
          0.0, 0.0, 0.0, 1.0, //
        ]) {
      return true;
    }
    return false;
  }
}

void main() {
  PatternMatching().report();
  EqualityChain().report();
}
```

</details>

I ran 96 in total:
- 3 methods
  - `getAsTranslation` (returning a `(double, double)` since we can't use `Offset` in non-Flutter contexts… yet 🙂)
  - `getAsScale`
  - `isIdentity`
- 4 variations for each method
- 8 runs per variation

I stored the results in a [public spreadsheet](https://docs.google.com/spreadsheets/d/182xC27hpV_ryqA4iElRgURpgcr2n1Oz0mwOnngIyHGI/edit?usp=sharing). There was a lot of back-and-forth, but overall, it looks like performance is almost exactly the same between the original and the pattern-matching versions.

But there was 1 outlier: the pattern-matching `isIdentity`'s performance was significantly worse overall than the original when most of the indices had to be checked before finding a mismatch. Perhaps it's much more difficult for pattern-matching to keep up with a regular conditional's performance if it can't "multitask" (i.e. assign a variable while it's checking the pattern).

<br><br>

Based on this information, I'm going to leave `isIdentity` the way it is, but use patterns for `getAsTranslation` and `getAsScale`.